### DISCUSSION TITLE
About torchmetrics Accuracy

### DISCUSSION THREAD

**Discussion by Aelees0807** [mrdbourke/pytorch-deep-learning#1374] [Q&A] [answered]
```
from torchmetrics import Accuracy

#setup metric
torchmetric_accuracy = Accuracy(task="multiclass" , num_classes = 4).to(device)

#Calculate accuracy
torchmetric_accuracy(y_preds , y_blob_test)
```

in the newer version we want to add "task" parameters ,  in our case the accuracy is 95% , but the `torchmetric_accuracy` is `tensor(0.3550, device='cuda:0')`

so what will be the reason behind it?

**Comment by yetval** [comment]
Hey, I looked into this. torchmetrics isn't the problem, when I give it the trained model's predictions it returns 0.9950, same as `accuracy_fn`.

A score of 0.3550 means the `y_preds` you passed in came from the untrained model. The notebook makes predictions once before training (in the logits to probabilities section) and once after, so if cells run out of order you end up scoring the untrained ones, which is basically random guessing for 4 classes.

Fix is to recompute predictions right before the metric:

```python
model_4.eval()
with torch.inference_mode():
    y_preds = torch.softmax(model_4(X_blob_test), dim=1).argmax(dim=1)

torchmetric_accuracy(y_preds, y_blob_test)
```

That should match your 95%.

**Comment by stefisha** [comment, marked as answer]
The issue is probably not `torchmetrics` itself, but the `y_preds` you are passing into it.

In the notebook, predictions are created before training and then again after training. If the cells were run out of order, `y_preds` may still contain predictions from the untrained model.

Try recalculating the predictions right before using `Accuracy`:

```python
model_4.eval()

with torch.inference_mode():
    y_preds = torch.softmax(model_4(X_blob_test), dim=1).argmax(dim=1)

torchmetric_accuracy = Accuracy(
    task="multiclass",
    num_classes=4
).to(device)

accuracy = torchmetric_accuracy(y_preds, y_blob_test)

print(accuracy)
```

This should give you a result much closer to the ~95% accuracy you were getting before.

Also make sure `y_preds` contains the predicted class labels like `0, 1, 2, 3`, rather than the raw logits.

If this resolves your question, please mark it as the answer. 🙂
